skimage/feature/sift.py

```diff
@@ -519,6 +519,8 @@ def _compute_orientation(self, positions_oct, scales_oct, sigmas_oct,
                     # result
                     ori = ((m + self._fit(hist[neigh]) + 0.5)
                            * 2 * np.pi / self.n_bins)
+                    if not np.isfinite(ori):
+                        continue
```

Comment:
Shouldn't the exception be handled explicitly (i.e., with a `try` statement and `except ZeroDivisionError`)? And within the `_fit` function?

Reply:
Let's say if we handle it by exception within `_fit`. What do you want to return by `_fit` when we catch the exception?

Reply:
I admit I would let it error, with a helpful message though: Maybe suggesting that the user pass `h` with a different precision, to circumvent the `ZeroDivisionError`?

Reply:
Hmm, sorry, I guess `h` is not really 'passed by the user' at this point; it's usually an intermediate result in another private function... Maybe try returning `h[0] - h[2]) / (2 * ((h[0] - h[1]) + (h[2] - h[1]))` as you wrote, and if this fails return `h[0] - h[2]) / (2 * (h[0] + h[2] - 2 * h[1])` as originally, so we have better chances overall...? And handle the exception in `_compute_orientation` instead?

Reply:
Sorry for the late reply. I was too busy to take a look on it.

> Maybe try returning h[0] - h[2]) / (2 * ((h[0] - h[1]) + (h[2] - h[1])) as you wrote, and if this fails return h[0] - h[2]) / (2 * (h[0] + h[2] - 2 * h[1])

I don't think it increases the chance. But, in any case, I guess it could be `h[0] == h[1] == h[2]` because 
```python
In [2]: ndi.maximum_filter([1., 1., 1.], [3], mode='wrap')
Out[2]: array([1., 1., 1.])
```
So, I think `ori` could be `nan` and `inf`, and `isfinite` can handle both cases.

> And handle the exception in _compute_orientation instead?

Actually, my PR description was wrong. The zero division was actually happening, but the error happened at [this line](https://github.com/scikit-image/scikit-image/blob/0df68491eedd3eb60f4d5e0fa7de670d3c2c934b/skimage/feature/sift.py#L547) with `ValueError: math domain error` because `angle` became `inf`.

So, unless we cast each element of `h` to python built-in type, we can't catch it as `ZeroDivisionError` because `h` is a numpy array.


By the way, I have two more PRs open for fixing bugs of SIFT (#7065 and #6862). So, it would be great if you can review them too.

Reply:
> Actually, my PR description was inaccurate.

No worries; feel free to edit your PR's title.

> Actually, my PR description was inaccurate. The zero division was actually happening, but the error happened at [this line](https://github.com/scikit-image/scikit-image/blob/0df68491eedd3eb60f4d5e0fa7de670d3c2c934b/skimage/feature/sift.py#L547) with ValueError: math domain error because angle became inf.

Thanks for investigating, @hayatoikoma. So, I think that your addition line 522 would be a good enough fix... How about reverting your change line 432?

> (Maybe better to replace cos and sin to numpy's function.)

I agree.

Reply:
Yes, it should handle both, but why do you want to revert line 432?

Reply:
Can you please add a test? Maybe you could write a toy version of your use case? So it breaks with current `main`, and passes with your proposed changes.

> Yes, it should handle both

Sorry, I don't get it: both what?

> but why do you want to revert line 432?

Let's not reverse the burden of proof... :wink: Unless you are fixing two different things and I didn't get it. I was under the impression that the change line 522 was enough because it handles the case when the angle becomes inf or nan. If you are changing line 432 as well, please add a commend and/or test for it. Thanks for following through!

Reply:
Thinking about it again, I thought I should just handle when the denominator becomes zero. Then, it should handle the cases I reported. What do you think?